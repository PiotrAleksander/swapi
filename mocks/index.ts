export const firstStarship = {
  name: "Y-wing",
  model: "BTL Y-wing",
  manufacturer: "Koensayr Manufacturing",
  cost_in_credits: "134999",
  length: "14",
  max_atmosphering_speed: "1000km",
  crew: "2",
  passengers: "0",
  cargo_capacity: "110",
  consumables: "1 week",
  hyperdrive_rating: "1.0",
  MGLT: "80",
  starship_class: "assault starfighter",
  pilots: [],
  films: [
    "https://swapi.co/api/films/2/",
    "https://swapi.co/api/films/3/",
    "https://swapi.co/api/films/1/"
  ],
  created: "2014-12-12T11:00:39.817000Z",
  edited: "2014-12-22T17:35:44.479706Z",
  url: "https://swapi.co/api/starships/11/"
};

export const secondStarship = {
  name: "X-wing",
  model: "T-65 X-wing",
  manufacturer: "Incom Corporation",
  cost_in_credits: "149999",
  length: "12.5",
  max_atmosphering_speed: "1050",
  crew: "1",
  passengers: "0",
  cargo_capacity: "110",
  consumables: "1 week",
  hyperdrive_rating: "1.0",
  MGLT: "100",
  starship_class: "Starfighter",
  pilots: [
    "https://swapi.co/api/people/1/",
    "https://swapi.co/api/people/9/",
    "https://swapi.co/api/people/18/",
    "https://swapi.co/api/people/19/"
  ],
  films: [
    "https://swapi.co/api/films/2/",
    "https://swapi.co/api/films/3/",
    "https://swapi.co/api/films/1/"
  ],
  created: "2014-12-12T11:19:05.340000Z",
  edited: "2014-12-22T17:35:44.491233Z",
  url: "https://swapi.co/api/starships/12/"
};
