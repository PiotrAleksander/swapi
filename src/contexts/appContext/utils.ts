import { ResourceType } from "../../types/enums/ResourceTypeEnum";
import api from "../../services/swapi";
import { Resource } from "../../types/Resource";
import {
  isPeopleResource,
  isStarshipResource,
  isNotFound
} from "../../types/typeguards";
import { Dispatch, SetStateAction } from "react";

export const fetchResources = async (resourceType: ResourceType) => {
  let firstIndex = Math.round(Math.random() * 10) + 1;
  let secondIndex = Math.round(Math.random() * 10) + 1;

  if (firstIndex === secondIndex) secondIndex += 1;

  const firstResourcePromise = api.fetchResource(resourceType, firstIndex);
  const secondResourcePromise = api.fetchResource(resourceType, secondIndex);
  const [firstResource, secondResource] = await Promise.all([
    firstResourcePromise,
    secondResourcePromise
  ]);
  return [firstResource, secondResource];
};

export const setWinnerAndCounter = (
  resources: Array<Resource>,
  setCounter: Dispatch<SetStateAction<number[]>>
): Array<Resource> => {
  let firstResource = resources[0];
  let secondResource = resources[1];
  if (isNotFound(firstResource) || isNotFound(secondResource)) {
    return resources;
  }

  if (isPeopleResource(firstResource) && isPeopleResource(secondResource)) {
    firstResource.winner =
      parseInt(firstResource.mass) > parseInt(secondResource.mass);
    secondResource.winner =
      parseInt(secondResource.mass) > parseInt(firstResource.mass);
  }

  if (isStarshipResource(firstResource) && isStarshipResource(secondResource)) {
    firstResource.winner =
      parseInt(firstResource.crew) > parseInt(secondResource.crew);
    secondResource.winner =
      parseInt(secondResource.crew) > parseInt(firstResource.crew);
  }

  if (firstResource.winner) {
    setCounter(prevCounter => [prevCounter[0] + 1, prevCounter[1]]);
  }

  if (secondResource.winner) {
    setCounter(prevCounter => [prevCounter[0], prevCounter[1] + 1]);
  }

  return [firstResource, secondResource];
};
