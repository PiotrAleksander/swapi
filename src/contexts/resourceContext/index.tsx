import React, { createContext } from "react";
import { IResourceContext } from "../../types/interfaces/IResourceContext";
import { ResourceType } from "../../types/enums/ResourceTypeEnum";
import { fetchResource } from "../../services/swapi";

export const ResourceContext = createContext<IResourceContext | null>(null);
