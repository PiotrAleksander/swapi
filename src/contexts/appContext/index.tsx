import React, {
  createContext,
  ReactNode,
  useState,
  useRef,
  useEffect,
  useCallback,
  Context
} from "react";
import { ResourceType } from "../../types/enums/ResourceTypeEnum";
import { IAppContext } from "../../types/interfaces/IAppContext";
import { fetchResources, setWinnerAndCounter } from "./utils";

const initialAppContext: IAppContext = {
  resourceType: ResourceType.PEOPLE,
  setResourceType: () => {},
  resources: [],
  counter: [0, 0],
  setCounter: () => {},
  onShuffle: () => {},
  isSending: false
};

let AppContext: Context<IAppContext>;
const { Provider, Consumer: AppConsumer } = (AppContext = createContext<
  IAppContext
>(initialAppContext));

const AppProvider = ({ children }: { children: ReactNode }) => {
  const [resourceType, setResourceType] = useState(ResourceType.PEOPLE);
  const [resources, setResources] = useState();
  const [counter, setCounter] = useState([0, 0]);
  const [isSending, setIsSending] = useState(false);
  const isMounted = useRef(true);

  useEffect(() => {
    return () => {
      // set unmounted on cleanup
      isMounted.current = false;
    };
  }, []);

  const onShuffle = useCallback(async () => {
    // don't send if sending already in progress
    if (isSending) return;
    setIsSending(true);
    try {
      // send the request
      const fetchedResources = await fetchResources(resourceType);

      const [firstResource, secondResource] = setWinnerAndCounter(
        fetchedResources,
        setCounter
      );

      setResources([firstResource, secondResource]);
    } catch (error) {
      setResources([]);
    } finally {
      if (isMounted.current)
        // only update if the provider is still mounted
        setIsSending(false);
    }
  }, [isSending, resourceType]); // update the callback if sending or resource type changes

  return (
    <Provider
      value={{
        resourceType,
        setResourceType,
        resources,
        counter,
        setCounter,
        onShuffle,
        isSending
      }}
    >
      {children}
    </Provider>
  );
};

export { AppProvider, AppConsumer };
export default AppContext;
