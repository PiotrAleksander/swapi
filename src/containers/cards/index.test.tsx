import React from "react";
import CardsContainer from ".";
import { render, fireEvent, waitForElement } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import { AppProvider, AppConsumer } from "../../contexts/appContext";
import { thirdStarship, fourthStarship } from "../../../mocks";

describe("CardsContainer", () => {
  let mockedFetch: jest.Mock;
  beforeEach(() => {
    mockedFetch = jest.fn();
    mockedFetch
      .mockResolvedValueOnce({
        json: async () => thirdStarship
      })
      .mockResolvedValueOnce({
        json: async () => fourthStarship
      });
    window.fetch = mockedFetch;
  });

  it("should render correctly", () => {
    const { container } = render(<CardsContainer />);
    expect(container).toMatchSnapshot();
  });

  it("should render if resources change", async () => {
    const { getByRole, getByText } = render(
      <AppProvider>
        <AppConsumer>
          {({ onShuffle }) => <button onClick={onShuffle}></button>}
        </AppConsumer>
        <CardsContainer />
      </AppProvider>
    );

    fireEvent.click(getByRole("button"));

    const firstHeading = await waitForElement(() =>
      getByText("TIE Advanced x1")
    );
    const secondHeading = await waitForElement(() => getByText("Executor"));

    expect(firstHeading).toBeInTheDocument();
    expect(secondHeading).toBeInTheDocument();
  });
});
