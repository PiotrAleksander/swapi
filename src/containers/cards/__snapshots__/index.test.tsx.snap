// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsContainer should render correctly 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-center"
      />
    </div>
  </div>
</div>
`;
